Add Base64Url support for netstandard 2.0

buyaa-n opened the pull request:
Add Base64Url package that supports netstandard 2.0 and .NET framework

Pretty much same implementation excluding the parts that not supported in netstandard like:
- No vectorization, though the serial part of the implementation kept as is
- No static abstract interface, therefore, no code sharing between Span<char>, Span<byte> overloads
- No `Span.IndexOfAnyExcept(SearchValues)` support so validation APIs logic differs around that are, but main logic is same  
- No Math.DivRem available

Fixes https://github.com/dotnet/runtime/issues/1658

buyaa-n committed:
Base64Url netstandard support package

ViktorHofer commented on src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:
```diff
@@ -0,0 +1,36 @@
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum);$(NetCoreAppCurrent)</TargetFrameworks>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <UseCompilerGeneratedDocXmlFile>false</UseCompilerGeneratedDocXmlFile>
```
Why disable the compiler generated doc xml file?

ericstj replied:
Yep - keep it enabled.  Best to use the source-authored docs here.

buyaa-n replied:
Sure, just copied from existing Microsoft.Bcl.* packages

ViktorHofer commented on src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:
```diff
@@ -0,0 +1,36 @@
+
+  <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
+  <PropertyGroup>
+    <IsPartialFacadeAssembly Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPartialFacadeAssembly>
```
I _think_ we should just include the "..ref/Microsoft.Bcl.Memory.Forwards.cs" file here instead of setting this property. @ericstj opinions?

ericstj replied:
Correct, it's better to hand author the forwards here.  No need to use partial facade generation.

ViktorHofer commented on src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
I understand that you follow the pattern from existing Bcl projects but I wonder why we need this reference source project and compile it, given that we don't ship it in the package. That seems like unnecessary work to me.

ericstj replied:
Seems reasonable to support a case where we don't have the ref project - but we'd want to make sure nothing else in the build system depends on it.

ViktorHofer replied:
Let's follow-up on that in a separate issue.

ViktorHofer committed:
Enable compiler doc generation and simplify forwards

ViktorHofer commented:
@buyaa-n I tried my feedback out in a code space. Here's the patch that you can directly apply:

```patch
From 2b35645c7eee21dfdbdd35855d709c9def195097 Mon Sep 17 00:00:00 2001
From: Viktor Hofer <viktor.hofer@microsoft.com>
Date: Tue, 18 Jun 2024 16:17:54 +0000
Subject: [PATCH] Enable compiler doc generation and simplify forwards

---
 .../src/Microsoft.Bcl.Memory.csproj                  | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj b/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
index aac0fba3d95..fa7bc6e2c33 100644
--- a/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
+++ b/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
@@ -3,7 +3,6 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum);$(NetCoreAppCurrent)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <UseCompilerGeneratedDocXmlFile>false</UseCompilerGeneratedDocXmlFile>
     <IsPackable>true</IsPackable>
     <!-- Disabling baseline validation since this is a brand new package.
          Once this package has shipped a stable version, the following line
@@ -19,17 +18,20 @@
 
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
-    <IsPartialFacadeAssembly Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPartialFacadeAssembly>
-    <OmitResources Condition="'$(IsPartialFacadeAssembly)' == 'true'">true</OmitResources>
+    <OmitResources Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</OmitResources>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+    <Compile Include="..\ref\Microsoft.Bcl.Memory.Forwards.cs" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
     <Compile Include="System\Buffers\Text\Base64UrlDecoder.cs" />
     <Compile Include="System\Buffers\Text\Base64UrlEncoder.cs" />
     <Compile Include="System\Buffers\Text\Base64UrlValidator.cs" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
     <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
   </ItemGroup>

```

ViktorHofer commented:
The allconfigurations leg is failing because the netstandard2.0 version of Base64Url has the ClsCompliant(false) attribute on it but the net9.0 version doesn't. If that's intentional, create a suppression file. The error explains how to do that.

ViktorHofer commented:
No, I don't have write permissions to your fork. You can either enable perms or apply the patch yourself (if `git apply ...` doesn't work, feel free to manually update the file).

buyaa-n commented:
> No, I don't have write permissions to your fork. You can either enable perms

For the PR `Allow edits and access to secrets by maintainers` checkbox is checked, I will try to enable perms on my fork 

> apply the patch yourself (if `git apply ...` doesn't work, feel free to manually update the file).

I tried that and `error: corrupt patch at line 45`

buyaa-n commented:
> No, I don't have write permissions to your fork. You can either enable perms

@ViktorHofer I added you as collaborator, please accept and apply your patch

buyaa-n committed:
Add supression for [ClsCompliant(false)] attribute

ViktorHofer commented:
The removal of the ClsCompliant(false) attribute is non-breaking, right? I'm 99% sure it isn't but just want to double check.

buyaa-n committed:
Remove ClsComplaint attribute and diff suppression

buyaa-n commented:
> The removal of the ClsCompliant(false) attribute is non-breaking, right? I'm 99% sure it isn't but just want to double check.

Weird there were ClsCompliant warning on `uint` `ushort` types that is why I added the attribute, but somehow now I don't see the warning when removing the attribute, now removed the attribute and suppression

ViktorHofer reviewed (approved):
Infrastructure / project changes LGTM

buyaa-n committed:
Use char type direcly instead of ushort, as we are not using vectorization

buyaa-n committed:
Fix bug in char overload and other small updates

stephentoub commented on src/libraries/Microsoft.Bcl.Memory/src/Resources/Strings.resx:
```diff
@@ -0,0 +1,126 @@
+    <value>Destination is too short.</value>
+  </data>
+  <data name="Format_BadBase64Char" xml:space="preserve">
+    <value>The input is not a valid Base64Url string as it contains a non Base64Url character, more than two padding characters, or an illegal character among the padding characters.</value>
```
```suggestion
    <value>The input is not a valid Base64Url string as it contains a non-Base64Url character, more than two padding characters, or an illegal character among the padding characters.</value>
```

stephentoub commented on src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:
```diff
@@ -0,0 +1,1343 @@
+                return ignoreWhiteSpace ?
+                    InvalidDataFallback(source, bytes, ref bytesConsumed, ref bytesWritten, isFinalBlock) :
+                    OperationStatus.InvalidData;
+            }
```
There's a lot of code being duplicated here. Is there no opportunity for further code sharing with the .NET 9 implementation?

buyaa-n replied:
Because cannot use static abstract methods in netstandard the main implementation cannot be shared, though I think some helper methods could be shared, assuming it will not need to add a netstandard target for CoreLib

stephentoub replied:
Would we be able to share it if we just used non-static abstract methods? e.g. instead have instance interface methods, implement those as instance methods on a struct, and pass in a default instance of the struct as an argument?

buyaa-n replied:
I guess it could, but how that would affect performance?

stephentoub replied:
I'd hope very little, but you'd need to actually measure to see.

buyaa-n replied:
OK, I'll give it try

stephentoub commented on src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:
```diff
@@ -0,0 +1,1343 @@
+            destination[2] = (byte)value;
+        }
+
+        private static bool IsValidPadding(uint padChar) => padChar == EncodingPadEqual || padChar == EncodingPadPercentage;
```
```suggestion
        private static bool IsValidPadding(uint padChar) => padChar is EncodingPadEqual or EncodingPadPercentage;
```
?

stephentoub commented on src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:
```diff
@@ -0,0 +1,1343 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsWhiteSpace(int value)
+        {
+            if (Environment.Is64BitProcess)
+            {
+                // For description see https://github.com/dotnet/runtime/blob/48e74187cb15386c29eedaa046a5ee2c7ddef161/src/libraries/Common/src/System/HexConverter.cs#L314-L330
+                // Lookup bit mask for "\t\n\r ".
+                const ulong MagicConstant = 0xC800010000000000UL;
+                ulong i = (uint)value - '\t';
+                ulong shift = MagicConstant << (int)i;
+                ulong mask = i - 64;
+                return (long)(shift & mask) < 0;
+            }
+
+            if (value < 32)
+            {
+                const int BitMask = (1 << (int)'\t') | (1 << (int)'\n') | (1 << (int)'\r');
+                return ((1 << value) & BitMask) != 0;
+            }
+
+            return value == 32;
```
This method will only ever be used with byte / char inputs, right? i.e. we don't need to care about int values > ushort.MaxValue? If so, I believe you could use this implementation, which is the 32-bit version of the 64-bit version used, and it would be appropriate for both 32-bit and 64-bit processes:
```C#
static bool IsWhiteSpace(int value)
{
    Debug.Assert(value >= 0 && value <= ushort.MaxValue);
    uint charMinusLowUInt32;
    return (int)((0xC8000100U << (short)(charMinusLowUInt32 = (ushort)(value - '\t'))) & (charMinusLowUInt32 - 32)) < 0;
}
```

buyaa-n replied:
Yes, it is only used for byte char inputs, thanks!

buyaa-n committed:
Apply suggestions from code review

Co-authored-by: Stephen Toub <stoub@microsoft.com>

buyaa-n committed:
Update src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

buyaa-n committed:
Use instance methods, share netstandard and .NET 9 code

buyaa-n committed:
Merge branch 'netstandard' of github.com:buyaa-n/runtime into netstandard

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs:
```diff
@@ -17,6 +19,10 @@ namespace System.Buffers.Text
         private const int MaxStackallocThreshold = 256;
+#pragma warning disable CA1805 // Member 's_base64UrlByteDecoder' is explicitly initialized to its default value
+        private static Base64UrlDecoderByte s_base64UrlByteDecoder = default;
+        private static Base64UrlDecoderChar s_base64UrlCharDecoder = default;
```
Having statics for these is a deoptimization. The call sites can just use eg `default(Base64UrlDecoderByte)`.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:
```diff
@@ -0,0 +1,1446 @@
+                    // This should never overflow since destLength here is less than int.MaxValue / 4 * 3 (i.e. 1610612733)
+                    // Therefore, (destLength / 3) * 4 will always be less than 2147483641
+                    Debug.Assert(destLength < (int.MaxValue / 4 * 3));
+#if NETCOREAPP
```
With the exception of the code in these ifdefs, and with calls being prefixed with `decoder.`, this whole file is an exact copy of what was in the other file, right? Or are there other changes?

buyaa-n replied:
Exactly as you said,, only changes are if-defs for netsandard, and type parameter to `decoder` no new changes. Well, there is small bugfix that related to the new tests added, nothing else new

Basically all internal members from `Base64Decoder` and `Base64Encoder` moved in this `Base64Herlper` type to avoid conflict referencing `Base64` from netstandard project between .NET 9 and netstandard versions.

stephentoub replied:
> Well, there is small bugfix that related to the new tests added, nothing else new

What is that fix?

buyaa-n replied:
I was fixed the bug found with Base64Url fuzzing only for byte overloads, the same bug left for char overloads.
 
The diff is here https://github.com/dotnet/runtime/pull/103617/commits/73acb51861f24f32ec1f2b3d2007486ca3ebbd42

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:
```diff
@@ -0,0 +1,1446 @@
+        [CompExactlyDependsOn(typeof(Avx512Vbmi))]
+        private static unsafe void Avx512Decode<TBase64Decoder, T>(TBase64Decoder decoder, ref T* srcBytes, ref byte* destBytes, T* srcEnd, int sourceLength, int destLength, T* srcStart, byte* destStart)
+     where TBase64Decoder : IBase64Decoder<T>
+     where T : unmanaged
```
nit: indenting

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:
```diff
@@ -0,0 +1,1446 @@
+#if NETCOREAPP
+                Base64.GetMaxDecodedFromUtf8Length(utf8Length);
+#else
+                0;
```
Why is this 0?

buyaa-n replied:
Well, this source will not called for netsandard,  though I think now after splitting the Base64 public vs internal sources I don't need to do this.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:
```diff
@@ -0,0 +1,1446 @@
+            public uint AdvSimdLutTwo3Uint1 => 0x1B1AFFFF;
+
+            public int GetMaxDecodedLength(int utf8Length) =>
+#if NETCOREAPP
```
Can we use `#if NET` instead of `#if NETCOREAPP` for all of these?

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs:
```diff
@@ -0,0 +1,802 @@
+    internal static partial class Base64Helper
+    {
+#pragma warning disable CA1805 // Member 's_base64ByteEncoder' is explicitly initialized to its default value
+        internal static Base64EncoderByte s_base64ByteEncoder = default;
```
As commented earlier, all of these statically cached instances can be removed, and just use `default(T)` instead of reading this field.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs:
```diff
@@ -0,0 +1,802 @@
+        [CompExactlyDependsOn(typeof(Avx512Vbmi))]
+        private static unsafe void Avx512Encode<TBase64Encoder, T>(TBase64Encoder encoder, ref byte* srcBytes, ref T* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, T* destStart)
+    where TBase64Encoder : IBase64Encoder<T>
+    where T : unmanaged
```
Nit: indenting

buyaa-n committed:
Merge conflicts

buyaa-n committed:
Apply feedbacks
